### PR TITLE
Json iterators

### PR DESCRIPTION
**Pull request by DmitriyMusatkin**
*Issue #, if available:*

*Description of changes:*

Json parser iterators for obj fields/array elements

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: include/aws/common/json.h [hunk 1]
```diff
@@ -166,6 +166,31 @@ bool aws_json_value_has_key(const struct aws_json_value *object, struct aws_byte
...
+ * iteration is sequential in order fields were initially parsed.
+ * @param object object to iterate over.
+ * @param on_key callback for when key is endcountered.
+ * @param user_data user data to pass back in callback.
+ * @return AWS_OP_SUCCESS when iteration finishes completely or exits early,
+ * AWS_OP_ERR if value is not an object.
+ */
+AWS_COMMON_API
+int aws_json_const_iterate_object(
+    struct aws_json_value *object,
```
Comment (graebm) [inline include/aws/common/json.h:190]: const object

File: include/aws/common/json.h [hunk 2]
```diff
@@ -166,6 +166,31 @@ bool aws_json_value_has_key(const struct aws_json_value *object, struct aws_byte
...
+    aws_json_on_key_encountered_const_fn *on_key,
```
Comment (graebm) [inline include/aws/common/json.h]: "on_member" probably a better name than "on_key" for the typedef and param.

according to json.org, and "object" is composed of "members"

File: include/aws/common/json.h [hunk 3]
```diff
@@ -211,6 +236,29 @@ size_t aws_json_get_array_size(const struct aws_json_value *array);
...
+ * @param array array to iterate over.
+ * @param on_key callback for when element is endcountered.
+ * @param user_data user data to pass back in callback.
+ * @return AWS_OP_SUCCESS when iteration finishes completely or exits early,
+ * AWS_OP_ERR if value is not an array.
+ */
+AWS_COMMON_API
+int aws_json_const_iterate_array(
+    struct aws_json_value *array,
+    aws_json_on_element_encountered_const_fn *on_element,
```
Comment (graebm) [inline include/aws/common/json.h:259]: "on_value" probably a better name than "on_element" for the typedef and param.

This is is more debatable than my comment on the object iterator.

While json.org does say an "array" is composed of "elements", an "element" is "ws value ws" and we don't super care about whitespace so 🤷‍♀️

Comment (DmitriyMusatkin) [reply include/aws/common/json.h:259]: I actually like on value here. reduces the cognitive load by not introducing another concept

File: include/aws/common/json.h [hunk 4]
```diff
@@ -166,6 +166,31 @@ bool aws_json_value_has_key(const struct aws_json_value *object, struct aws_byte
  */
 AWS_COMMON_API
 int aws_json_value_remove_from_object(struct aws_json_value *object, struct aws_byte_cursor key);
+
+/**
+ * @brief callback for iterating keys of an object
+ * return false to exit early
+ */
+typedef bool(aws_json_on_member_encountered_const_fn)(
```
Comment (graebm) [inline include/aws/common/json.h:174]: trivial: "iterating keys" -> "iterating members"

File: include/aws/common/json.h [hunk 5]
```diff
@@ -166,6 +166,31 @@ bool aws_json_value_has_key(const struct aws_json_value *object, struct aws_byte
...
+    const struct aws_byte_cursor *key,
+    const struct aws_json_value *value,
+    void *user_data);
+
+/**
+ * @brief iterates though fields of the object.
```
Comment (graebm) [inline include/aws/common/json.h:180]: trivial: "iterates though fields" -> "iterates through members"

also, decent spell checker plugin for vscode: https://marketplace.visualstudio.com/items?itemName=streetsidesoftware.code-spell-checker

File: include/aws/common/json.h [hunk 6]
```diff
@@ -166,6 +166,31 @@ bool aws_json_value_has_key(const struct aws_json_value *object, struct aws_byte
...
+ * iteration is sequential in order fields were initially parsed.
+ * @param object object to iterate over.
+ * @param on_memeber callback for when member is endcountered.
```
Comment (graebm) [inline include/aws/common/json.h:183]: memeber endcountered

File: include/aws/common/json.h [hunk 7]
```diff
@@ -166,6 +166,31 @@ bool aws_json_value_has_key(const struct aws_json_value *object, struct aws_byte
...
+ * return false to exit early
+ */
+typedef bool(aws_json_on_member_encountered_const_fn)(
+    const struct aws_byte_cursor *key,
+    const struct aws_json_value *value,
+    void *user_data);
+
+/**
+ * @brief iterates though fields of the object.
+ * iteration is sequential in order fields were initially parsed.
```
Comment (graebm) [inline include/aws/common/json.h:181]: I wouldn't bother giving details that might change if we swapped out JSON implementations

File: include/aws/common/json.h [hunk 8]
```diff
@@ -211,6 +236,29 @@ size_t aws_json_get_array_size(const struct aws_json_value *array);
  */
 AWS_COMMON_API
 int aws_json_value_remove_array_element(struct aws_json_value *array, size_t index);
+
+/**
+ * @brief callback for iterating elements of an array.
```
Comment (graebm) [inline include/aws/common/json.h:241]: trivial "iterating elements" -> "iterating values"

File: include/aws/common/json.h [hunk 9]
```diff
@@ -211,6 +236,29 @@ size_t aws_json_get_array_size(const struct aws_json_value *array);
...
+ * return false to exit early.
+ */
+typedef bool(aws_json_on_value_encountered_const_fn)(size_t index, const struct aws_json_value *value, void *user_data);
+
+/**
+ * @brief iterates though elements of an array.
```
Comment (graebm) [inline include/aws/common/json.h:247]: trivial: "elements" -> "values"

File: include/aws/common/json.h [hunk 10]
```diff
@@ -211,6 +236,29 @@ size_t aws_json_get_array_size(const struct aws_json_value *array);
...
+ * iteration is sequential starting with 0th element.
+ * likely to be faster than manually calling get element at.
+ * @param array array to iterate over.
+ * @param on_value callback for when element is endcountered.
```
Comment (graebm) [inline include/aws/common/json.h:251]: "element" -> "value"
endcountered -> encountered

**Review by graebm — APPROVED** [review]

**Pull request merged**